### PR TITLE
x86: support very early printk() if desired

### PR DESCRIPTION
**Pull request by tyrantlizard77**
Adapted from similar code in the x86_64 port.
Useful when debugging boot problems on actual x86
hardware if a JTAG isn't handy or feasible.

Turn this on for qemu_x86, to help with CI logs if
something breaks early on.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

### CODE REVIEW

**Review by dcpleung — APPROVED** [review]

**Comment by dcpleung** [comment]
Some UARTs are done with MMIO so this may not work. Might be good to put a note somewhere that this requires UART on I/O ports.

File: arch/x86/core/early_serial.c [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+{
+	unsigned char ret;
+
+	__asm__ volatile("inb %1, %0;\n\t" : "=a"(ret) : "d"(port));
+	return ret;
+}
+
+static void serout(int c)
+{
+	while (!(ioport_in8(_PORT + 5) & 0x20)) {
```
Comment (gnuless) [inline arch/x86/core/early_serial.c]: I hate the naked constants here.  In the future maybe we could move this into the ns16550 driver file (just as some #ifdef'd functions that are like "here's the standalone stuff for early console") so we can use the definitions there, plus keep everything 16550-related in one place.

File: arch/x86/core/early_serial.c [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	ioport_out8(_PORT+1, 0);    /* IER = 0 */
+	ioport_out8(_PORT+3, 0x80); /* LCR = 8n1 + DLAB select */
+	ioport_out8(_PORT,   1);    /* Divisor Latch low byte */
+	ioport_out8(_PORT+1, 0);    /* Divisor Latch high byte */
+	ioport_out8(_PORT+3, 0x03); /* LCR = 8n1 + DLAB off */
+	ioport_out8(_PORT+4, 0x03); /* MCR = DTR & RTS asserted */
+
+	/* Will be replaced later when a real serial driver comes up */
+	__stdout_hook_install(console_out);
+	__printk_hook_install(console_out);
```
Comment (andyross) [inline arch/x86/core/early_serial.c]: Doesn't stdout go to printk() by default?  I swear it did that already.  If not, it probably should.

**Review by andyross — APPROVED** [review]
Thanks.  This is definitely a lot earlier.  That said, it's not as early as it was.  On x86_64, the port initialization happens at the very top of the 32 bit stub (cstart(), in xuk-stub32.c, which is reached after a 7-instruction header from the bootloader, where it's also used for logging within the stub).  This version has to wait until Zephyr z_cstart() calls back into kernel_arch_init(), which is actually after the log subsystem (but obviously not the log backends) gets initialized.

The way the OS integration worked originally is that the arch layer (or alternatively anywhere) defines a z_arch_printk_char_out function. This symbol is the default value of the printk _char_out handler, and there is a weak noop version available.  There is no "registration" process, you don't need to have any part of the OS running.  It's a promise from the lower layer(s) that "This function can be called in any context to emit a byte to a useful console".  The original version was a hook to the esp32 console firmware, which worked like this.

I'd strongly suggest using that mechanism instead, and if necessary lazy-initiale the qemu serial port.  That would avoid the problem of having to wait for kernel_arch_init().  It also plays better with other users of the same trick, like for example a future ACRN console logger, etc...  You'd just select one of them with a kconfig and wouldn't have to juggle all the initialization hooks.

But this is still a big advantage over waiting for the device init layers to start up.

**Review by ghost — APPROVED** [review]
Couple of comments, no show-stoppers.  

Arguably could be moved to the board/SoC level instead of the arch, but I'm not fussed.

**Comment by andrewboie** [comment]
> which is actually after the log subsystem (but obviously not the log backends) gets initialized.

I have no idea why the log subsystem feels it needs to init that early, before the architecture has initialized itself, this seems a mistake. I think it should be moved after, in which case we are equivalent since everything before this is in assembly / not calling printk

> The way the OS integration worked originally is that the arch layer (or alternatively anywhere) defines a z_arch_printk_char_out function. This symbol is the default value of the printk _char_out handler, and there is a weak noop version available. There is no "registration" process, you don't need to have any part of the OS running. 

I have no idea what you mean by this, since all this depends on is the code in lib/os/printk.c which uses no OS services other than whatever char_out does

The "registration" you are whining about is:

```
void __printk_hook_install(int (*fn)(int))
{
	_char_out = fn;
}
```

Who cares?? This is existing code, it works, I'm going to use it. It really baffles me that you decided to re-implement printk in x86_64 instead of just using what's already there.

> It's a promise from the lower layer(s) that "This function can be called in any context to emit a byte to a useful console". The original version was a hook to the esp32 console firmware, which worked like this.

We're not running on xtensa and this is hardcoded to a specific serial port.
I would rather not make this any more complex than it already is.

File: arch/x86/core/early_serial.c (hunk 1 above)
Comment (andrewboie) [reply arch/x86/core/early_serial.c]: I'll punt on 16550 driver unification, but I agree about naked constants, will fix

File: arch/x86/core/early_serial.c [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+/* Super-primitive 8250/16550 serial output-only driver, 115200 8n1 */
+
+#define _PORT DT_UART_NS16550_PORT_0_BASE_ADDR
+
+static inline void ioport_out8(unsigned short port, unsigned char b)
+{
+	__asm__ volatile("outb %0, %1;\n\t" : : "a"(b), "d"(port));
+}
+
+static inline unsigned char ioport_in8(unsigned short port)
```
[the remark this answers is not in the document]
Comment (andrewboie) [reply arch/x86/core/early_serial.c]: derp will do

File: arch/x86/core/early_serial.c (hunk 2 above)
Comment (andrewboie) [reply arch/x86/core/early_serial.c]: think this is only for printf, i'll remove it

**Comment by andrewboie** [comment]
> 
> 
> Some UARTs are done with MMIO so this may not work. Might be good to put a note somewhere that this requires UART on I/O ports.

@dcpleung added to Kconfig help text.

**Pull request merged**